### DISCUSSION TITLE
Support for Channel Resolution Greater Than 10-bit

### DISCUSSION THREAD

**Discussion by arunh** [ExpressLRS/ExpressLRS#3055] [Ideas]
ELRS currently supports 10-bit channel resolution. This works well for serial connections between a receiver and flight controller. When a mechanical linkage is introduced such as on a fixed-wing aerobatic model, I think there are situations where a higher-resolution is of benefit.

It is rare to be able to get a mechanical servo linkage able to use the entire servo range, often due to physical constraints out of the model designers control. Even when this is possible, there are scenarios where a pilot might want as little as (+/-15%) of full throw over the range of the transmitter stick. A real world example of this is a fixed-wing freestyle model where mechanical throws to suit both 3D (+/-120%) and precision flight (+/-15%) are required on the same mechanical linkage.

With a 10-bit/1024-pos resolution, a (+/-15%) throw gives just 128 unique positions which is noticeable in terms of feel and precision with a good quality 12-bit/4096-pos servo. Improving the channel resolution beyond 10-bits would improve this situation and bring ELRS in line with other existing high-end systems such as Jeti in terms of resolution whilst conferring the benefits of a LoRa based system. The overall effect here should be improved performance for PWM setups with mechanical servo-driven linkages across a wide range of use-cases.

This request is to make the resolution used to send channels configurable by the end-user up to [13-bits](https://github.com/crsf-wg/crsf/wiki/CRSF_FRAMETYPE_SUBSET_RC_CHANNELS_PACKED) (maximum supported by CRSF). Additional to this would be the ability to control the number of channels sent at this resolution as the number needed varies with different model types. Controlling the number of channels sent would allow some control over the channel update rate.

**Comment by Fourflies** [comment]
@arunh Just thinking out loud, would it work if a PWM receiver remapped the 10-bit range into a narrow PWM range that the receiver sends out to the servo?

**Reply by arunh** [reply]
If we assume that the TX mixer can compute pulse width to a resolution of 1μs (it could be more accurate), there are 1231 possible positions in the range 885μs to 2115 μs currently used by ELRS / CRSF Extended Limits. As such 10bits results in some different mixer values sharing the same 10-bit channel value. 11bits can provide a unique representation of this data. As such, I think it is necessary to have an 11bit+ data path from the transmitter.

**Comment by Divefire** [comment]
To second this, it would be really useful to have that higher resolution available for surface model use. The more precision available on the main control channels the better for surface models.

**Reply by Divefire** [reply]
And to add a note to myself and anyone  looking, there is an 11-bit branch  but it's behind a long way. Does serve as a proof of concept at least and a good starting point?

https://github.com/ExpressLRS/ExpressLRS/compare/master...11bit-mod
